docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,247 @@
+`switch` to check whether the message contains an incoming IBC packet, and if so
+verify the Merkle proof.
+
+```go
+// Pseudocode
+type ProofVerificationDecorator struct {
+  clientKeeper ClientKeeper
+  channelKeeper ChannelKeeper
+}
+
+func (pvr ProofVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  var flag bool
+
+  var portID, channelID string
+
+  for _, msg := range tx.GetMsgs() {
+    var err error
+    switch msg := msg.(type) {
+    case client.MsgUpdateClient:
+      err = pvr.clientKeeper.UpdateClient(msg.ClientID, msg.Header)
+    case channel.MsgPacket:
+      err = pvr.channelKeeper.VerifyPacket(msg.Packet, msg.Proofs, msg.ProofHeight)
+      if flag {
+        if msg.PortID != portID || msg.channelID != channelID {
+          return ctx, errors.New("Transaction cannot include IBC packets from different ports")
+        }
+        portID = msg.PortID
+        channelID = msg.ChannelID
+      }
+      flag = true
+      // Store the empty acknowledgement for convinience
+      pvr.channelKeeper.SetPacketAcknowledgement(ctx, msg.PortID, msg.ChannelID, msg.Sequence, []byte{})
+    case chanel.MsgAcknowledgement:
+      err = pvr.channelKeeper.VerifyAcknowledgement(msg.Acknowledgement, msg.Proof, msg.ProofHeight)
+      if flag {
+        if msg.PortID != portID || msg.channelID != channelID {
+          return ctx, errors.New("Transaction cannot include IBC packets from different ports")
+        }
+        portID = msg.PortID
+        channelID = msg.ChannelID
+      }
+      flag = true
+    case channel.MsgTimeoutPacket:
+      err = pvr.channelKeeper.VerifyTimeout(msg.Packet, msg.Proof, msg.ProofHeight, msg.NextSequenceRecv)
+      if flag {
+        if msg.PortID != portID || msg.channelID != channelID {
+          return ctx, errors.New("Transaction cannot include IBC packets from different ports")
+        }
+        portID = msg.PortID
+        channelID = msg.ChannelID
+      }
+      flag = true
+    default:
+      if flag {
+        return ctx, errors.New("Transaction cannot include both IBC packet messasges and normal messages")
+      }
+      continue
+    }
+
+    if err != nil {
+      return ctx, err
+    }
+  }
+  return next(ctx, tx, simulate)
+}
+```
+
+Where `MsgUpdateClient`, `MsgPacket`, `MsgAcknowledgement`, `MsgTimeoutPacket`
+are `sdk.Msg` types correspond to `handleUpdateClient`, `handleRecvPacket`,
+`handleAcknowledgementPacket`, `handleTimeoutPacket` of the routing module,
+respectively.
+
+An attacker can insert a failiing message before any of packet receiving message
+preventing the packet messages to be processed but keeping the sequence increased.
+If a transaction contains a packet receiving messages, any possibly failing 
+messages should not be included in the transaction.
+
+The `ProofVerificationDecorator` will be inserted to the top level application.
```

Comment:
I think there needs to be an explanation here for why the above logic is necessary to implement at the ante-handler level rather than just doing everything on the handler level.